### PR TITLE
Add parsing from xContent to SearchResponse

### PR DESCRIPTION
**Pull request by cbuescher**
In preparation to be able to parse SearchResponse from its rest representation
for the java rest client, this adds fromXContent to SearchResponse. Most of the
information in the original object is preserved when parsing it back. However,
the exceptions in the "failure" section won't be identical to the original ones
on the server side since they are parsed back to a generic
ElasticsearchException on the receiving side. Also the "aggregations", "suggest"
and "profile" section parsing is currently skipped and will be added by
subsequent PRs.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: just an idea, but maybe we could instead generate random response without exceptions and check that like we usually do, and have specific tests for exceptions? Otherwise exceptions affect the way we test the rest of the response which doesn't sound optimal.

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: you could call Strings.toString instead I think

File: core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: leftover?

File: core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: s/renderes/rendered . Instead, we could build what we expect given the exception? Isn't it just wrapping what we have into an ElasticsearchException with the same message?

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: do we still need these changes?

**Review by javanna — CHANGES_REQUESTED** [review]
left some comments, looks good though. I would start a new feature branch based on this given that the parsing will be completed only once we have also done profile, suggest and aggs sections which will take a while.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: I think it needs another `else` block with `throwUnknownField(currentFieldName, parser.getTokenLocation())` here

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: Can we use `failures.toArray(new ShardSearchFailure[failures.size()])` ?

File: core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java]: I think in other classes we didn't add a `_FIELD` suffix

File: core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java]: `XContentParserUtils.throwUnknownToken()` instead?

File: core/src/main/java/org/elasticsearch/rest/action/RestActions.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/rest/action/RestActions.java]: I think it would make sense, yes.

File: core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: That would be nice to revert this change (it "pollutes" a bit the Git history)

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: Thanks for the indentation ;)

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
Comment (tlrx) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: Do we need this empty path?

**Review by tlrx — CHANGES_REQUESTED** [review]
I like it, thanks for doing it @cbuescher. I think it makes sense to have a dedicated branch until every pieces are merged in, though you could already add ShardSearchFailure changes in its own pull request.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: right, changes that

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: I think this is still needed as its own class because it is used as an "incomplete" search response in e.g. SeachPhaseController#merge() and in some other places where some information we need for the SearchResponse ctor is not available yet. I also wanted to keep the amount of changes small, so I'd prefer to leave it for now. This would be another refactoring that is not necessary for parsing.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: It could, it still has the read/write methods. However, they should not be used anywhere else than in SearchResponse, so I'd go one step in the other direction and even reduce the visibility of write/read.

File: core/src/main/java/org/elasticsearch/rest/action/RestActions.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/rest/action/RestActions.java]: sure, I will remove it

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: yes, already notices that with your comment above

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: yep

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: Good to know.

File: core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: yep

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: I added these path arguments to be able to better debug errors in map-comparisons. When we compare xContent and some nested map has different size, we currently only get the expected and actual size as output. This is very useful to print in the test error message, so I'd like to leave it.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: Right, on that level its more like an unexpected token, we have another helper for that which I will use.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: Sure, good point.

File: core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java]: I just added this to the other class in this PR, I think it better communicates that this is a field name.

File: core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java]: Yes, forgot about that for a second

File: core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: makes sense, thanks for catching, didn't see that

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: As mentioned in another comment, I added these path arguments to be able to better debug errors in map-comparisons when we compare xContent. This empty path is used as initial value, assertMapEquals() will be called recursively later.

File: core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java]: Good point, I will change this. However, I'd like the createTestItem() method to still be able to output an object with set "failures", so I will make this controllable from the outside. In our "normal" fromXContent() test we can use xContent comparison again then, In another test with "failures" at the moment the only thing we can really check is that we can parse the response without throwing an error I think. Other more detailed tests are covered by ShardSearchFailureTests already I think.

File: core/src/main/java/org/elasticsearch/rest/action/RestActions.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/rest/action/RestActions.java]: This is a good idea, I started looking into it but it looks like this needs quiet a bit of moving code around, so I'd prefer to do that in a separate PR if you're okay with that.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: that's fine.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: ok because although the class can be created separately, it should always be used as part of the SearchResponse when serializing?

File: core/src/main/java/org/elasticsearch/rest/action/RestActions.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/action/RestActions.java]: fine

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: I don't follow, there is a new argument to these methods that always gets passed in as "" ? or what am I missing?

Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: I got it, ok!

Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: In assertToXContentEquivalent() we start with an empty path. Whenever assertMapEquals() finds an expected key, we append that to the path. Whenever an expected value or a map size differs, we additionally print the collected path information up until then. Its basically a recursion, "path" being the collector. the assertListEquals() method also needs the argument because we need to pass it on for other nested objects inside.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: shall we use -1 as default values, just to make sure not to confuse them with actual values that get returned?

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: shall timedOut also be a Boolean? maybe not because it's always returned while terminatedEarly is not? trying to understand the difference between the two

File: core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: Instead, can we build what we expect given the exception? Isn't it just wrapping what we have into an ElasticsearchException with the same message? Or does it get too complicated? I think Tanguy did something similar in some other test.

**Review by javanna — APPROVED** [review]
left a couple of comments, LGTM otherwise, no need for another review round on my end

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: yes, thats it. "timedOut" is always set, always rendered out, so it should always be parsed. "terminatedEarly" can also be "null" in the original object, it only rendered when its present, so its okay if we don't parse it.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: done

File: core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: I will look into it but won't spend too much time on it at this point.

Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: Building the "expected" value (either the object or its xContent rendering) is really difficult and will break easiliy without using ElasticsearchException#fromXContent(), which unfortunately is what we want to test here. The way this wraps the original exception and rewrites the message in it is hard to mock without actually repeating it, which defies the usefulness of a test IMHO. Still looking but I'm a bit clueless at this point.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: I just checked where InternalSearchResponse is used, all places that I found where this is instantiated it is later wrapped in a SearchResponse in the same method, so I think it's never sent across the wire independently.

File: core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: It can be complicated to rebuild the object, how about doing something like:
```
assertEquals(parsed.getCause().getMessage(), "Elasticsearch exception [type=parse_exception, reason=" + originalMsg +"]");
```

Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java]: Yes, thats another slightly different option, that works when we don't test for nested causes which I think we don't have to here. Will change this, although it doesn't add much to the current version.

**Comment by cbuescher** [comment]
@javanna @tlrx  since we agreed to put this on a a separate branch I opened #22699 than can be merged to master already. I will rebase this PR as well for now until we open the feature branch

**Comment by cbuescher** [comment]
Closing this since it is quiet stale now.

**Pull request closed without merging**
